src/main/java/org/broadinstitute/hellbender/tools/exome/sexgenotyper/ContigPloidyAnnotationTableReader.java

```diff
@@ -0,0 +1,131 @@
+
+        final ContigClass contigClass;
+        final String contigClassString = dataLine.get(ContigPloidyAnnotationTableColumn.CONTIG_CLASS);
+        if (!ContigClass.CONTIG_CLASS_NAMES_SET.contains(contigClassString)) {
+            throw new UserException.BadInput("Bad contig class: provided value: " + contigClassString + ", acceptable values: " +
+                    ContigClass.CONTIG_CLASS_NAMES_SET.stream().collect(Collectors.joining(", ", "[", "]")));
+        } else {
+            contigClass = ContigClass.valueOf(contigClassString);
+        }
+
+        final Map<String, Integer> ploidyMap = new HashMap<>();
+        try {
+            /* all lines must all ploidy annotations defined in the header */
+            ploidyTagsSet.forEach(tag -> ploidyMap.put(tag, dataLine.getInt(tag)));
+        } catch (final IllegalArgumentException ex) {
+            throw new UserException.BadInput("All lines in the contig ploidy annotation table must have values for all" +
+                    " ploidy classes; " + ex.getMessage());
+        }
+
+        return new ContigPloidyAnnotation(contigName, contigClass, ploidyMap);
+    }
+
+    /**
+     * Read contig ploidy annotations from file.
+     *
+     * @param contigPloidyAnnotationsFile a file
+     * @return a list of contig ploidy annotations
+     * @throws IOException if a read error occurs
+     */
+    public static List<ContigPloidyAnnotation> readContigPloidyAnnotationsFromFile(@Nonnull final File contigPloidyAnnotationsFile)
+            throws IOException {
+        Utils.regularReadableUserFile(contigPloidyAnnotationsFile);
+        try {
+            return readContigPloidyAnnotationsFromReader(contigPloidyAnnotationsFile.getAbsolutePath(),
+                    new FileReader(contigPloidyAnnotationsFile));
+        } catch (final FileNotFoundException ex) {
+            throw new UserException.CouldNotReadInputFile("Could not read contig ploidy annotations file " +
+                    contigPloidyAnnotationsFile.getAbsolutePath());
+        }
+    }
+
+    /**
+     * Read contig ploidy annotations from an abstract reader.
+     *
+     * @param contigAnnotationSourceName a string identifier for the reader.
+     * @param contigAnnotationReader an instance of {@link Reader}.
+     * @return list of contig ploidy annotations
+     * @throws IOException if a read error occurs
+     */
+    public static List<ContigPloidyAnnotation> readContigPloidyAnnotationsFromReader(@Nonnull final String contigAnnotationSourceName,
+                                                                                     @Nonnull final Reader contigAnnotationReader)
+            throws IOException {
+        /* read contig annotations */
+        try (final ContigPloidyAnnotationTableReader reader =
+                     new ContigPloidyAnnotationTableReader(contigAnnotationSourceName, contigAnnotationReader)) {
+            return reader.stream().collect(Collectors.toList());
```

Comment:
Sure. `ContigPloidyAnnotationTableReader` extends `TableReader<ContigPloidyAnnotation>`. Any subclass of `TableReader<T>` inherits the method `public Stream<T> stream()` which is a stream of type `T`, i.e. the data type the table reader generates from reading records. One can make a `Stream<T>` out of any class that implements `Iterable<T>` via instantiating a `Spliterator<T>` from the `iterator()`. A spliterator is an iterator that can split into two iterators. Think of streams as (possibly multiple) iterators that sweep the container (possibly asynchronously and in parallel). Take a look at `TableReader`. It is also useful to reader about iterators, spliterators, and streams from a java book (I put one on your desk!)